### PR TITLE
Ask confirmation when closing a document

### LINKED ISSUE #13258: Ask for confirmation when closing a tab
### Problem

Some users want to get a confirmation dialog to show up when they close a tab.

> Currently a confirmation dialog is opened only if the document is in a dirty state.

### Proposed Solution

Add an option to always display a close confirmation dialog.

### Additional context

N/A

### PR DESCRIPTION
**Pull request by alec-kr**
## References
Fixes #13258.

## Code changes

- Add a new settings to ask document closing confirmation (`false` by default  - this is an opt-in feature).
- Prompt the user to confirm closing the widget
- Add a optional check box to change the settings to not ask for confirmation
- Ask confirmation and save or confirmation and discard if the file is dirty
- Add `stateChange` signal in `DocumentManager` and `DocumentWidgetManager` to propagate the change in an attribute to be able to update the settings.

confirm-close-doc

### CODE REVIEW

File: packages/docmanager/src/widgetmanager.ts [hunk 1]
```diff
@@ -312,6 +312,10 @@
           await context.saveAs();
         }
       }
+
+      if (!shouldClose) {
```
[the remark this answers is not in the document]
Comment (jtpio) [reply packages/docmanager/src/widgetmanager.ts:316]: Looks like the warning is relevant, the `if (!shouldClose)` check is within the `if (shouldClose)` check.

[the remark this answers is not in the document]
Comment (alec-kr) [reply packages/docmanager/src/widgetmanager.ts]: @jtpio No idea how I missed that. I have moved the ```if``` statement now.

**Comment by fcollonval** [comment]
Thanks @alec-kr for starting this. Could you add the code following the step by step given in https://github.com/jupyterlab/jupyterlab/issues/13258#issuecomment-1282397003

**Comment by alec-kr** [comment]
I will not be able to solve this issue anytime soon, due to some unforeseen issues.

I just want to make it clear that I am no longer working on the issue, and it is available to anyone who may want to contribute.

File: packages/docmanager-extension/schema/plugin.json [hunk 2]
```diff
@@ -122,6 +122,12 @@
       "description": "Length of save interval in seconds",
       "default": 120
     },
+    "confirmClosingDocument": {
+      "type": "boolean",
+      "title": "Ask confirmation to close a document",
+      "description": "Whether to ask confirmation to close a document or not.",
```
Comment (krassowski) [inline packages/docmanager-extension/schema/plugin.json:128]:
```suggestion
      "title": "Ask for confirmation to close a document",
      "description": "Whether to ask for confirmation to close a document or not.",
```

**Comment by krassowski** [comment]
This seems to be doing the opposite of feature requested in #13258 (allowing to opt-out from conditional confirmation dialog rather than allowing to opt-in into unconditional confirmation dialog), do I read this right?

**Comment by fcollonval** [comment]
> This seems to be doing the opposite of feature requested in #13258 (allowing to opt-out from conditional confirmation dialog rather than allowing to opt-in into unconditional confirmation dialog), do I read this right?

I went for a opt-out option to make it discoverable by users. As most users don't look at changelog, this will allow them to see the new feature at least once. But I'm happy to change the default settings if it is found more appropriate.

**Comment by jtpio** [comment]
> I went for a opt-out option to make it discoverable by users. As most users don't look at changelog, this will allow them to see the new feature at least once. But I'm happy to change the default settings if it is found more appropriate.

Some folks would also prefer this to be opt-out by default to avoid showing another popup.

Maybe this could still be opt-in for both master / 4.0 an 3.6 for now as suggested in https://github.com/jupyterlab/jupyterlab/issues/13258? And reconsidered as opt-out for 4.0.

**Comment by jtpio** [comment]
This could also be discussed at the weekly dev meeting to gauge the overall feeling in case some folks haven't had time to look at this PR.

**Comment by fcollonval** [comment]
Following the discussion here and at the weekly call, 4bb116e makes this feature opt-in.

**Comment by fcollonval** [comment]
bot please update galata snapshots

**Comment by fcollonval** [comment]
@meeseeksdev please backport to 3.6.x

**Comment by jtpio** [comment]
Just checking if this change is intended since it's not directly related to this PR?

image

**Comment by fcollonval** [comment]
It is a side effect of some galata improvements. As this is for the benchmark test, it makes sense to have less noise.

**Comment by jtpio** [comment]
OK sounds good thanks :+1:

**Pull request merged**